[cmake] Auto-complete via clangd auto-setup v2

ntrrgc opened the pull request:
#### 4b85cbe26eff585fe8586af43bd7199f326b0caf
<pre>
[cmake] Auto-complete via clangd auto-setup v2
<a href="https://bugs.webkit.org/show_bug.cgi?id=274909">https://bugs.webkit.org/show_bug.cgi?id=274909</a>

Reviewed by Elliott Williams.

Re-upload of the patch including a fix for the internal Apple builds
that do not contain a Tools/ directory. See
<a href="https://bugs.webkit.org/show_bug.cgi?id=274909#c7">https://bugs.webkit.org/show_bug.cgi?id=274909#c7</a> for details.

Additionally, this version has been modified to not depend on PyYAML or
Jinja2 and use simpler stdlib replacements instead, as Apple internal
builds cannot use any non-builtin Python packages.

---

 # Goals

This patch aims to bring out-of-the box completion with clangd for
WebKit developers using CMake builds.

clangd is an open source language server for providing code-completion,
go-to-definition and similar smart features for C++ code by analyzing it
with the same backend clang uses. **You do not need to compile WebKit
with clang to take advantage of this**, as clang strives to be flag
compatible with GCC, and even MSVC through `clang-cl`.

Using clangd with WebKit was already possible before, but required
manual intervention in the project files to set it up.

This patch attempts to clear some of the existing pitfalls,
particularly for CMake builds, so that more people take advantage of
Clangd without having to put conscious effort into troubleshooting.

This patch also makes WebKit stock clangd setup compatible with both
Mac and non-Mac platforms and allows developers to customize clangd
configuration without dirtying their repository.

---

Getting clangd right takes a lot of context and moving pieces, so this
is one of those very long commit messages. Sections with (NEW) in the
title explain changes introduced by this patch, whereas the rest hope to
explain the background and why those changes are necessary or useful.

 # Architecture (NEW)

Several scripts have been added in Tools/clangd. A diagram describing
the build system targets, dependencies and products is also included.

The clangd auto-setup needs to generate the following artifacts:
 * .clangd configuration file
 * compile_commands.json file/symlink in the WebKit repo root.

Starting with this patch `.clangd` is no longer committed in the
repository it is now .gitignore&apos;d and auto-generated during the build.

 # Compatibility with XCode builds (NEW)

This patch doesn&apos;t improve or worsen the status of clangd support in
XCode builds.

XCode build users are still expected to generate compile_commands.json
manually as per the wiki states:
<a href="https://trac.webkit.org/wiki/Clangd#macOS1">https://trac.webkit.org/wiki/Clangd#macOS1</a>

    make release EXPORT_COMPILE_COMMANDS=YES
    generate-compile-commands WebKitBuild/Release

The generation of the `.clangd` file is also included in the XCode
build, as part of the WTF XCode project.

 # clangd config generation (NEW)

The script `update-clangd-config` autogenerates a `.clangd` file with an
«autogenerated» marker at the top. The file will only be updated as long
as that marker exists, therefore allowing end-user customization if
needed.

The `.clangd` file is generated from a template, which will produce
slightly different options for Mac and non-Mac platforms, fixing the
regression introduced in <a href="https://github.com/WebKit/WebKit/pull/23273.">https://github.com/WebKit/WebKit/pull/23273.</a>
The template previously used Jinja2, but in this version of the patch
uses the more rudimentary stdlib&apos;s strings.Template to avoid depending
on external libraries.

`update-clangd-config` is made part of the build by default. For XCode
builds, it is generated as part of WTF. For CMake builds, it is enabled
as a default target as long as CLANGD_AUTO_SETUP is set, which is the
default when building inside the WebKitBuild directory.

 # clangd configuration changes (NEW)

This patch also makes several changes to the clangd configuration file
to fix issues detected during testing:

 * UnifiedSource files are no longer indexed.
 * The `-fcoroutines` flag is stripped, as it is not recognized by some
   versions of clang, and by extension, clangd.
 * `-xobjective-c++` is now only passed in macOS to avoid depending on
   macOS-only headers in other systems.
 * `config.h` is no longer prepended to files within the ThirdParty
   directory.

 # compile_commands.json generation: context

CMake already generates a compile_commands.json file when the Ninja or
Makefile file is generated as long as CMAKE_EXPORT_COMPILE_COMMANDS is
ON, which is already the default in DEVELOPER_MODE.

When using unified builds (which is the default), that
`compile_commands.json` is somewhat subpar, as the language server is
unaware of most of the source .cpp files and hence has to require to
heuristics to match them to other build commands.

On the CMake side, a RewriteCompileCommands target exists that parses
that file along with unified sources to generate an improved file at
`&lt;BUILD_DIR&gt;/DeveloperTools/compile_commands.json`. Unfortunately, this
may not be obvious to many developers, which may be unaware of its
existence.

On the XCode builds side, this responsibility would fall under the
`generate-compile-commands` script, but it&apos;s seemingly not currently
implemented. Fixing that shouldn&apos;t be a difficult change but it&apos;s
outside of the scope of this patch. For more details:
<a href="https://github.com/WebKit/WebKit/pull/30784#issuecomment-2264933253">https://github.com/WebKit/WebKit/pull/30784#issuecomment-2264933253</a>

 # Same compile_commands.json path for non-unified builds (NEW)

The first change this patch makes is solving the unified builds vs
non-unified builds dichotomy by introducing a new build target for
non-unified builds named SymlinkCompileCommandsInDeveloperTools.

This way, **no matter the type of build**,
`&lt;BUILD_DIR&gt;/DeveloperTools/compile_commands.json` is the path to be
relied upon.

 # compile_commands.json vs builds

For a `compile_commands.json` file to be used, it needs to exist in a
parent directory of the sources being worked on.

Since any generated `compile_commands.json` is specific for a given build
(e.g. Debug vs Release), with build-specific flags, developers need to
make a choice of what build to use for the `compile_commands.json` in
their WebKit checkout.

In XCode builds, this is passed explicitly to the
generate-compile-commands command.

In CMake builds, this is usually accomplished by symlinking the file
into the root of the WebKit repository.

Developers who frequently switch and remove builds need to be careful to
not let the compile_commands.json symlink become broken, as that would
cause the language server to rely on much worsened heuristics.

 # compile_commands.json symlink autogeneration and management (NEW)

This patch introduces automatic compile_commands.json symlink handling.
A new target is introduced that intentionally runs with every build:
`UpdateCompileCommandsSymlink`, which relies on the
`update-compile-commands-symlink` script, also introduced by this patch.
The script is meant to be fast enough to not introduce any perceivable
delay in builds.

`update-compile-commands-symlink` receives as arguments a symlink path
and a configuration file listing a series of candidate paths. The
symlink is updated to point to the first candidate path that exists.
In the previous version of this patch, that file was in YAML format, but
this version uses Python&apos;s ConfigParse INI syntax to avoid depending on
any external libraries.

This list of candidate paths is declared in a new .gitignore&apos;d
`update-compile-commands-symlink.conf` file that is generated early
during any CMake build that uses `CLANGD_AUTO_SETUP`.

`update-compile-commands-symlink.conf` will be automatically updated
from `Tools/clangd/update-compile-commands-symlink.conf.example`; but
similarly to `.clangd`, a user can disable this and manage the file on
their own, in this case by setting the `user_managed` property inside
the file to True.

 # Testing

clangd auto-setup has been tested with a Linux x86_64 machine using
webkit-container-sdk and Visual Studio Code, both with the default GCC
and with clang 16, with a fresh installation of VS Code (no settings).

It is possible to confirm if a file passes under clangd by using the
`clangd --check` command along with the compile lines used during
compilation, which can be queried from compile_commands.json.

A couple barebones scripts showing this usage have been provided in
Tools/clangd as documentation: `clangd-check-file` and
`clangd-check-codebase`.

I was able to run clangd-check-codebase with very few .cpp files
failing, and the failing files had good reasons for failing, like
missing includes that were obscured by unified builds.

Canonical link: <a href="https://commits.webkit.org/282175@main">https://commits.webkit.org/282175@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6f61716370020b70f34faafb74820b556d3feb8a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60892 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40251 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13468 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64823 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11439 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63022 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47927 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11714 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49234 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7937 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62926 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37594 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52758 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30060 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34160 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9981 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10352 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/53994 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55985 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10277 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66552 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/60139 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4836 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10102 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56597 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4857 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52720 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56790 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13850 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4009 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/81894 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36056 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14271 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37138 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38231 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36883 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple reviewed (changes requested):
Thanks for the ping!

`update-clangd-config` using an auto-installed jinja2 will unfortunately be an issue for us, as we need to be able to build WebKit in environments where there is no network access.

One thing I'm thinking about is, does the clangd generator really need to be this complex?  If I understand correctly, the only thing that is dynamically generated is the `-xobjective-c++` argument when building on a Mac.

What if we had multiple clangd files (`Tools/clangd/clangd-Darwin.yaml`, `Tools/clangd/clangd-Linux.yaml`, etc.), and all the build-time script did was symlink the correct one into the top-level? It could even be done with a shell one-liner like:

    ln -sfh Tools/clang/clangd-$(uname -s).yaml "${WK_WORKSPACE_DIR}/.clang"

emw-apple commented on .gitignore:
```diff
@@ -4,6 +4,8 @@
 .directory
 /WebKitBuild/
+/.clangd
+/update-compile-commands-symlink.conf
```
Can this go in a `Tools/clangd/.gitignore` file? Doesn't seem like it needs to be stored in the top level directory.

ntrrgc replied:
As far I know, not really, as those ignore patterns would only apply inside `Tools/clangd`.

I tested doing the change just in case, and it indeed seem to only apply to `Tools/clangd` and not to the root of the repository, which sounds reasonable to me.

emw-apple reviewed (approved):
The bits that touch the Xcode build look fine to me. I'd recommend getting a second review from someone who knows the Linux + CMake build better.

TingPing reviewed (commented):
CMake changes LGTM

ntrrgc committed:
[cmake] Auto-complete via clangd auto-setup v2
https://bugs.webkit.org/show_bug.cgi?id=274909

Reviewed by Elliott Williams.

Re-upload of the patch including a fix for the internal Apple builds
that do not contain a Tools/ directory. See
https://bugs.webkit.org/show_bug.cgi?id=274909#c7 for details.

Additionally, this version has been modified to not depend on PyYAML or
Jinja2 and use simpler stdlib replacements instead, as Apple internal
builds cannot use any non-builtin Python packages.

---

 # Goals

This patch aims to bring out-of-the box completion with clangd for
WebKit developers using CMake builds.

clangd is an open source language server for providing code-completion,
go-to-definition and similar smart features for C++ code by analyzing it
with the same backend clang uses. **You do not need to compile WebKit
with clang to take advantage of this**, as clang strives to be flag
compatible with GCC, and even MSVC through `clang-cl`.

Using clangd with WebKit was already possible before, but required
manual intervention in the project files to set it up.

This patch attempts to clear some of the existing pitfalls,
particularly for CMake builds, so that more people take advantage of
Clangd without having to put conscious effort into troubleshooting.

This patch also makes WebKit stock clangd setup compatible with both
Mac and non-Mac platforms and allows developers to customize clangd
configuration without dirtying their repository.

---

Getting clangd right takes a lot of context and moving pieces, so this
is one of those very long commit messages. Sections with (NEW) in the
title explain changes introduced by this patch, whereas the rest hope to
explain the background and why those changes are necessary or useful.

 # Architecture (NEW)

Several scripts have been added in Tools/clangd. A diagram describing
the build system targets, dependencies and products is also included.

The clangd auto-setup needs to generate the following artifacts:
 * .clangd configuration file
 * compile_commands.json file/symlink in the WebKit repo root.

Starting with this patch `.clangd` is no longer committed in the
repository it is now .gitignore'd and auto-generated during the build.

 # Compatibility with XCode builds (NEW)

This patch doesn't improve or worsen the status of clangd support in
XCode builds.

XCode build users are still expected to generate compile_commands.json
manually as per the wiki states:
https://trac.webkit.org/wiki/Clangd#macOS1

    make release EXPORT_COMPILE_COMMANDS=YES
    generate-compile-commands WebKitBuild/Release

The generation of the `.clangd` file is also included in the XCode
build, as part of the WTF XCode project.

 # clangd config generation (NEW)

The script `update-clangd-config` autogenerates a `.clangd` file with an
«autogenerated» marker at the top. The file will only be updated as long
as that marker exists, therefore allowing end-user customization if
needed.

The `.clangd` file is generated from a template, which will produce
slightly different options for Mac and non-Mac platforms, fixing the
regression introduced in https://github.com/WebKit/WebKit/pull/23273.
The template previously used Jinja2, but in this version of the patch
uses the more rudimentary stdlib's strings.Template to avoid depending
on external libraries.

`update-clangd-config` is made part of the build by default. For XCode
builds, it is generated as part of WTF. For CMake builds, it is enabled
as a default target as long as CLANGD_AUTO_SETUP is set, which is the
default when building inside the WebKitBuild directory.

 # clangd configuration changes (NEW)

This patch also makes several changes to the clangd configuration file
to fix issues detected during testing:

 * UnifiedSource files are no longer indexed.
 * The `-fcoroutines` flag is stripped, as it is not recognized by some
   versions of clang, and by extension, clangd.
 * `-xobjective-c++` is now only passed in macOS to avoid depending on
   macOS-only headers in other systems.
 * `config.h` is no longer prepended to files within the ThirdParty
   directory.

 # compile_commands.json generation: context

CMake already generates a compile_commands.json file when the Ninja or
Makefile file is generated as long as CMAKE_EXPORT_COMPILE_COMMANDS is
ON, which is already the default in DEVELOPER_MODE.

When using unified builds (which is the default), that
`compile_commands.json` is somewhat subpar, as the language server is
unaware of most of the source .cpp files and hence has to require to
heuristics to match them to other build commands.

On the CMake side, a RewriteCompileCommands target exists that parses
that file along with unified sources to generate an improved file at
`<BUILD_DIR>/DeveloperTools/compile_commands.json`. Unfortunately, this
may not be obvious to many developers, which may be unaware of its
existence.

On the XCode builds side, this responsibility would fall under the
`generate-compile-commands` script, but it's seemingly not currently
implemented. Fixing that shouldn't be a difficult change but it's
outside of the scope of this patch. For more details:
https://github.com/WebKit/WebKit/pull/30784#issuecomment-2264933253

 # Same compile_commands.json path for non-unified builds (NEW)

The first change this patch makes is solving the unified builds vs
non-unified builds dichotomy by introducing a new build target for
non-unified builds named SymlinkCompileCommandsInDeveloperTools.

This way, **no matter the type of build**,
`<BUILD_DIR>/DeveloperTools/compile_commands.json` is the path to be
relied upon.

 # compile_commands.json vs builds

For a `compile_commands.json` file to be used, it needs to exist in a
parent directory of the sources being worked on.

Since any generated `compile_commands.json` is specific for a given build
(e.g. Debug vs Release), with build-specific flags, developers need to
make a choice of what build to use for the `compile_commands.json` in
their WebKit checkout.

In XCode builds, this is passed explicitly to the
generate-compile-commands command.

In CMake builds, this is usually accomplished by symlinking the file
into the root of the WebKit repository.

Developers who frequently switch and remove builds need to be careful to
not let the compile_commands.json symlink become broken, as that would
cause the language server to rely on much worsened heuristics.

 # compile_commands.json symlink autogeneration and management (NEW)

This patch introduces automatic compile_commands.json symlink handling.
A new target is introduced that intentionally runs with every build:
`UpdateCompileCommandsSymlink`, which relies on the
`update-compile-commands-symlink` script, also introduced by this patch.
The script is meant to be fast enough to not introduce any perceivable
delay in builds.

`update-compile-commands-symlink` receives as arguments a symlink path
and a configuration file listing a series of candidate paths. The
symlink is updated to point to the first candidate path that exists.
In the previous version of this patch, that file was in YAML format, but
this version uses Python's ConfigParse INI syntax to avoid depending on
any external libraries.

This list of candidate paths is declared in a new .gitignore'd
`update-compile-commands-symlink.conf` file that is generated early
during any CMake build that uses `CLANGD_AUTO_SETUP`.

`update-compile-commands-symlink.conf` will be automatically updated
from `Tools/clangd/update-compile-commands-symlink.conf.example`; but
similarly to `.clangd`, a user can disable this and manage the file on
their own, in this case by setting the `user_managed` property inside
the file to True.

 # Testing

clangd auto-setup has been tested with a Linux x86_64 machine using
webkit-container-sdk and Visual Studio Code, both with the default GCC
and with clang 16, with a fresh installation of VS Code (no settings).

It is possible to confirm if a file passes under clangd by using the
`clangd --check` command along with the compile lines used during
compilation, which can be queried from compile_commands.json.

A couple barebones scripts showing this usage have been provided in
Tools/clangd as documentation: `clangd-check-file` and
`clangd-check-codebase`.

I was able to run clangd-check-codebase with very few .cpp files
failing, and the failing files had good reasons for failing, like
missing includes that were obscured by unified builds.

Canonical link: https://commits.webkit.org/282175@main
